From 4f63ba151b73d3108cabf5e1c717f76c8e428d45 Mon Sep 17 00:00:00 2001
From: GaryOderNichts <garyodernichts@gmail.com>
Date: Sun, 18 Dec 2022 14:13:14 +0100
Subject: [PATCH] Fix XPutBackEvent() to support GenericEvents

With commit d6d6cba90215d323567fef13d6565756c9956f60 generic events could no longer be put back, since _XEventToWire would always return _XUnknownNativeEvent for them.
---
 src/PutBEvent.c | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/src/PutBEvent.c b/src/PutBEvent.c
index f7b74b31..b8273e1a 100644
--- a/src/PutBEvent.c
+++ b/src/PutBEvent.c
@@ -85,15 +85,20 @@ XPutBackEvent (
 	int type = event->type & 0177;
 
 	LockDisplay(dpy);
-	fp = dpy->wire_vec[type];
-	if (fp == NULL)
-		fp = _XEventToWire;
-	ret = (*fp)(dpy, event, &wire);
-	if (ret)
+	if (type == GenericEvent)
+		ret = _XPutBackEvent(dpy, event);
+	else
 	{
-		ret = (*dpy->event_vec[type])(dpy, &lib, &wire);
+		fp = dpy->wire_vec[type];
+		if (fp == NULL)
+			fp = _XEventToWire;
+		ret = (*fp)(dpy, event, &wire);
 		if (ret)
-			ret = _XPutBackEvent(dpy, &lib);
+		{
+			ret = (*dpy->event_vec[type])(dpy, &lib, &wire);
+			if (ret)
+				ret = _XPutBackEvent(dpy, &lib);
+		}
 	}
 	UnlockDisplay(dpy);
 	return ret;
-- 
GitLab

